### DISCUSSION TITLE
Source generator additional files without using `<AdditionalFiles>`

### DISCUSSION THREAD

**Discussion by pr8x** [dotnet/roslyn#66978] [Q&A] [answered]
Is there a way to feed additional files into a source generator without using `<AdditionalFiles>` in my csproj?

**Comment by jasonmalinowski** [comment, marked as answer]
There needs to be some hint at the MSBuild level, because even MSBuild needs to know about the inputs. If you're packaging your source generator as a NuGet package, you can do the MSBuild stuff in there though without having to update the project file directly.

**Reply by pr8x** [reply]
Ok. The reason I am asking is that MSBuild seems to rebuild projects if the included file is missing. 

> Potential build performance issue in 'XY.csproj'. The project does not appear up-to-date after a successful build: Input AdditionalFiles item 'FOLDER' does not exist and is required, not up-to-date

It's weird (possibly a bug?) that it treats missing input files as "not up-to-date". I tried working around this by adding `Condition="Exists('FOLDER')"` to the attribute, but that doesn't help.

**Reply by sharwell** [reply]
Here's an example of a conditional include:

```xml
  <ItemGroup>
    <!-- Show launchSettings.json in the project if it exists. -->
    <None Include="$(AppDesignerFolder)\launchSettings.json" Condition="Exists('$(AppDesignerFolder)\launchSettings.json')" />
  </ItemGroup>
```

**Reply by pr8x** [reply]
Ah okay, I thought I have to use `AdditionalFiles`. Will check this. Thanks

**Reply by sharwell** [reply]
You do need to use `<AdditionalFiles>`. I'm just showing a concrete example of how the `Condition` relates to the `Include` (it'll have the exact same content).

**Reply by pr8x** [reply]
I am currently doing this:

```
<AdditionalFiles Include="Assets">
      <Condition>Exists('Assets')</Condition>
</AdditionalFiles>
```

However, this doesn't seem tor work. It still rebuilds the project. (Note: Assets is a folder in the same directory as csproj file)

**Reply by sharwell** [reply]
You can't pass a folder to `<AdditionalFiles>`. It's looking for a file named `Assets`.

**Reply by sharwell** [reply]
You could use a wildcard:

```xml
<AdditionalFiles Include="Assets\**\*.*" />
```

I'm not sure if that can be simplified to `Assets\**\*` and/or `Assets\**`, but something in there should work.

Wildcard expansions don't need `Condition` on existence because they only expand to items that exist.

**Reply by pr8x** [reply]
Ah, I see. I will try that.

**Reply by pr8x** [reply]
That seems to work well. Thank you!
